RFC: bluetooth: audio: Fix NULL dereference

finikorg opened the pull request:
Fixes NULL dereferences

finikorg committed:
bluetooth: audio: Fix NULL dereference

Fixes NULL dereference if ep == NULL and codec is not NULL and we
dereference ep after the changed line.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

Thalley reviewed (changes requested):
This will break stuff if `ep == NULL` and `codec != NULL`, as we dereference `ep` after the changed line. 

Can you elaborate on what the problem is? Should be more explicit in the commit message.

finikorg committed:
bluetooth: audio: stream: Fix NULL dereference

Fix condition for the loop, making sure that stream->ep is not NULL
since it gets dereferenced after this code.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
> This will break stuff if `ep == NULL` and `codec != NULL`, as we dereference `ep` after the changed line.
> 
> Can you elaborate on what the problem is? Should be more explicit in the commit message.

Fixed commit message

Thalley commented on subsys/bluetooth/audio/stream.c:
```diff
@@ -477,7 +477,7 @@ static void audio_stream_qos_cleanup(const struct bt_conn *conn,
 
 	SYS_SLIST_FOR_EACH_CONTAINER(&group->streams, stream, _node) {
-		if (stream->conn != conn && stream->ep != NULL) {
+		if (stream->conn != conn || stream->ep == NULL) {
```
This change does somewhat make sense, but currently in the code there is no place where `stream->ep` will be NULL when this function is called. 

Can you please elaborate on what issue you are attempting to fix?

finikorg replied:
Warnings with static tool. If `stream->ep` cannot be NULL the NULL check should be removed.

Thalley replied:
Which tool did you use, and can you post the warning here?? Just so we have a reference for the future :)

Anyways, we could replace the `stream->ep != NULL` with an `__ASSERT(stream->ep != NULL` so that the warning gets fixed, but we still have some check here in case that we would otherwise introduce a bug somewhere. 

Also please remove your first commit in this PR, as that isn't a valid change.

finikorg replied:
The warning is obvious, if there is NULL check then it can be NULL. If it cannot be NULL please remove NULL check.
The same applies to the first patch.

Thalley replied:
The reason why I am asking for the tool and warning, is that there are so many static code analyzers out there, and some of they are even mutually exclusive (i.e. some require changes that would make others fail), and thus we cannot adhere to all of them.

finikorg commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -1206,7 +1206,7 @@ static int ascs_ep_set_codec(struct bt_audio_ep *ep, uint8_t id, uint16_t cid,
 
-	if (ep == NULL && codec == NULL) {
+	if (ep == NULL) {
 		return -EINVAL;
```
@Thalley so you can guarantee that if ep == NULL then codec also == NULL? Otherwise we can have NULL dereference.

Thalley replied:
I assume you mean that if `ep != NULL` then `codec != NULL`? 

But yes, since `ascs_ep_set_codec` is only called one place, using the address of statically allocated values (which can not be NULL). In fact, in the current tree, `ep` can never be NULL.

finikorg replied:
No I mean what I wrote. If ep == NULL and codec != NULL we have NULL dereference. If ep cannot be NULL please remove then NULL check.
